### ISSUE TITLE
Docker client should perform monolithic uploads for small layers

### ISSUE THREAD

**Issue by ngorskig** [docker/docker#17841]
Currently the docker client always performs a streaming upload (POST, PATCH, PUT). This works well for larger images (presumably where the time it takes to read an image and compress it is long enough for parallel network work to be useful), but for smaller images this is inefficient. See `graph/push_v2.go:pushV2Image` for the area of interest.

In my prototype (https://github.com/ngorskig/docker/tree/prototype/monolithicuploads), I found that performing a monolithic upload was faster for images up to 16MB. For images <=1MB, monolithic uploads are 2-3 times faster. Monolithic uploads also have lower variance (an order of magnitude in my tests).

The linked prototype contains two commits. One commit changes `pushV2Image` to perform a branch on image size; if the image is "small" (<=16MB), the compressed image is loaded into memory and we call `BlobIngester.Put` instead of starting a streaming upload.

The other commit changes `BlobIngester.Put` to perform a monolithic POST. For prototyping purposes, it is in the vendored docker/distribution code; getting the client to do this will need a prerequisite change in docker/distribution to be merged into `vendor/`. This is the first blocking issue.

Additionally, the distribution registry doesn't currently support monolithic uploads. Support will need to be added for the client to work with the registry, but luckily this seems pretty trivial. I have a prototype of this as well (https://github.com/ngorskig/distribution/tree/prototype/monolithicuploads). This is the second blocking issue.

I will file two issues in docker/distribution for these two blockers and link to them here. Let me know what you think.

**Comment by aaronlehmann** [comment]
@ngorskig: Do you have any theories on why monolithic uploads are faster for small files? I'd like to better understand the performance bottleneck before changing the way uploads work. There may be things making the current upload process suboptimal that we can fix without a cutover to a different way of uploading small files.

**Comment by unclejack** [comment]
Uploading tiny layers in parallel might be a better idea.

**Comment by stevvooe** [comment]
@unclejack 
> Uploading tiny layers in parallel might be a better idea.

Agreed. The bottleneck is currently parallelization.

**Comment by ngorskig** [comment]
Monolithic uploads will be faster when the latency of network is greater than the latency of compression. If compression takes long enough, streaming partial results in parallel will be faster. For small images, compression is less than network time so it's faster to compress and send it all in one go. Networks are slow, so this threshold will capture a non-trivial amount of images (our data suggests 90%+).

Parallel uploads will be another big improvement, but it's orthogonal to faster small-image uploads.

Just to keep things data-driven, here's some analysis with DockerHub over 1000 samples:

**Time to create a streaming upload:**
Mean: 0.94897564 seconds
99th percentile: 1.61520218 seconds

**Time to stream the layer:**
Mean: 0.68932064 seconds
99th percentile: 1.42322060 seconds

**Time to commit the layer:**
Mean: 3.8393425 seconds
99th percentile: 11.0105302 seconds

As a comparison to monolithic uploads, I captured the time it takes to PUT the manifest. The manifest was 1391 bytes, and so I made sure my layer was also 1391 bytes. (This was the uncompressed size for the layer; since it's random data, the actual streamed size would be slightly more. Close enough.)

**Time to upload manifest:**
Mean: 1.0568439 seconds
99th percentile: 2.82009838 seconds

The complete layer upload span:

**Time from create to commit for a layer:**
Mean: 5.4778799 seconds
99th percentile: 12.5234034 seconds

I got these numbers by adding some logs statements in the daemon, then extracting the timestamps and doing a bit of math. See: https://github.com/ngorskig/docker/commit/7ea554a3abcfed48474e8a2856d004a2966525d1.

This isn't exactly the same performance characteristic as a real monolithic upload (clearly the commit step does some work server-side), but it's the general idea. Compressing a layer this small takes about a tenth of a second, so spending 1-2 seconds telling the server "here I come!" in parallel doesn't help. 

I tested the time it takes to compress the image until it was at least 1 second, and unsurprisingly it was around 16 MB on my machine. Once it takes longer to compress the image, sharding off data to the server in the background makes sense.

For reference, my average ping time to https://registry-1.docker.io/v2/ is 0.27583 seconds.

**Comment by stevvooe** [comment]
@ngorskig Thanks for the analysis.

Obviously, these numbers are dominated by s3 round trip time. The hard optimization would be to move registry metadata out of s3 and only store data. This would get the upload time easily under 1s without changing the protocol.

As point of interest, a monolithic is going to be dominated by commit time, so monolithic uploads may still be in the range of 11s, even with this optimization. To avoid that, make sure to use the digest parameter to short circuit the upload.

Please submit your plan in #1170 and we can work from there.

**Comment by sunkuet02** [comment]
Is it possible, to set the docker engine always to upload files in Monolithic approach rather than Chunk approach ?

**Comment by stevvooe** [comment]
@sunkuet02 No, and doing so won't likely help performance.

**Comment by stevvooe** [comment]
@sunkuet02 No. docker/distribution#1170 was a proposal to implement it.

Ideally, we should address round trips in the registry, as that is the cause of latency here.
